# (3주차) 9월28일
> 파이토치를 이용하여 회귀모형 학습하기 (2) 

- toc:true
- branch: master
- badges: true
- comments: true
- author: 최규빈

### 강의영상 

> youtube: https://youtube.com/playlist?list=PLQqh36zP38-xuwoiidMPSXtBmrzKA9uqH

`-` (1/3): 9월14-16일 강의노트의 일부내용 추가설명 

`-` (2/3): torch.nn.Linear()를 사용하여 yhat을 계산하기, torch.nn.MSELoss()를 이용하여 loss를 계산하기 

`-` (3/3): 과제설명 

### Import 

In [1]:
import torch 
import numpy as np 

### Data

`-` model: $y_i= w_0+w_1 x_i +\epsilon_i = 2.5 + 4x_i +\epsilon_i, \quad i=1,2,\dots,n$ 

`-` model: ${\bf y}={\bf X}{\bf W} +\boldsymbol{\epsilon}$

- ${\bf y}=\begin{bmatrix} y_1 \\ y_2 \\ \dots \\ y_n\end{bmatrix}, \quad {\bf X}=\begin{bmatrix} 1 & x_1 \\ 1 & x_2 \\ \dots \\ 1 & x_n\end{bmatrix}, \quad {\bf W}=\begin{bmatrix} 2.5 \\ 4 \end{bmatrix}, \quad \boldsymbol{\epsilon}= \begin{bmatrix} \epsilon_1 \\ \dots \\ \epsilon_n\end{bmatrix}$

In [2]:
torch.manual_seed(43052)
n=100
ones= torch.ones(n)
x,_ = torch.randn(n).sort()
X = torch.vstack([ones,x]).T
W = torch.tensor([2.5,4])
ϵ = torch.randn(n)*0.5
y = X@W + ϵ
ytrue = X@W

### 이전방법요약 

`-` step1: yhat 

`-` step2: loss 

`-` step3: derivation 미분

`-` step4: update 

### step1: yhat

`-` feedforward 신경망을 설계하는 과정

`-` 이 단계가 잘 완료되었다면, 임의의 ${\bf\hat{W}}$을 넣었을 때 $\bf\hat{y}$를 계산할 수 있어야 함 

#### 방법1: 직접선언 (내가 공식을 알고 있어야 한다)

In [3]:
What=torch.tensor([-5.0,10.0],requires_grad=True)

In [4]:
yhat1=X@What

In [5]:
yhat1

tensor([-29.8211, -28.6215, -24.9730, -21.2394, -19.7919, -19.6354, -19.5093,
        -19.4352, -18.7223, -18.0793, -16.9040, -16.0918, -16.0536, -15.8746,
        -14.4690, -14.3193, -13.6426, -12.8578, -12.5486, -12.4213, -11.9484,
        -11.1034, -10.8296, -10.6210, -10.5064, -10.0578,  -9.8063,  -9.7380,
         -9.7097,  -9.6756,  -8.8736,  -8.7195,  -8.6880,  -8.1592,  -7.7752,
         -7.7716,  -7.7339,  -7.7208,  -7.6677,  -7.1551,  -7.0004,  -6.8163,
         -6.7081,  -6.5655,  -6.4480,  -6.3612,  -6.0566,  -5.6031,  -5.5589,
         -5.2137,  -4.3446,  -4.3165,  -3.8047,  -3.5801,  -3.4793,  -3.4325,
         -2.3545,  -2.3440,  -1.8434,  -1.7799,  -1.5386,  -1.0161,  -0.8103,
          0.4426,   0.5794,   0.9125,   1.1483,   1.4687,   1.4690,   1.5234,
          1.6738,   2.0592,   2.1414,   2.8221,   3.1536,   3.6682,   4.2907,
          4.8037,   4.8531,   4.9414,   5.3757,   5.3926,   5.6973,   6.0239,
          6.1261,   6.5317,   7.2891,   8.4032,   8.4936,   9.27

#### 방법2: torch.nn.Linear() 사용

In [6]:
net = torch.nn.Linear(in_features=2 ,out_features=1, bias=False) 
# in_features : 입력 차원( n은 생략 )
# out_features : 출력 차원( n은 생략 )
# bias : 상수항 삽입 유무 
## 처음에 가정한 모델이 X의 차원이 nx2 이므로 이미 bias term(상수항)이 포함된 상태이다.
## 만약 X의 차원을 nx1로 가정했다면 bias 를 삽입해야한다.

In [7]:
net.weight.data
# torch에서 제공하는 초기 변수값 

tensor([[0.3320, 0.1982]])

In [8]:
net.weight.data=torch.tensor([[-5.0,10.0]])
# 초기 변수값을 원하는 tensor 값으로 변경 

In [9]:
net.weight.data

tensor([[-5., 10.]])

In [10]:
net(X)
# 결과값이 100x1 의 행렬로 출력됨.

tensor([[-29.8211],
        [-28.6215],
        [-24.9730],
        [-21.2394],
        [-19.7919],
        [-19.6354],
        [-19.5093],
        [-19.4352],
        [-18.7223],
        [-18.0793],
        [-16.9040],
        [-16.0918],
        [-16.0536],
        [-15.8746],
        [-14.4690],
        [-14.3193],
        [-13.6426],
        [-12.8578],
        [-12.5486],
        [-12.4213],
        [-11.9484],
        [-11.1034],
        [-10.8296],
        [-10.6210],
        [-10.5064],
        [-10.0578],
        [ -9.8063],
        [ -9.7380],
        [ -9.7097],
        [ -9.6756],
        [ -8.8736],
        [ -8.7195],
        [ -8.6880],
        [ -8.1592],
        [ -7.7752],
        [ -7.7716],
        [ -7.7339],
        [ -7.7208],
        [ -7.6677],
        [ -7.1551],
        [ -7.0004],
        [ -6.8163],
        [ -6.7081],
        [ -6.5655],
        [ -6.4480],
        [ -6.3612],
        [ -6.0566],
        [ -5.6031],
        [ -5.5589],
        [ -5.2137],


In [11]:
yhat2=net(X)

#### 방법3: torch.nn.Linear()사용, bias=True

In [12]:
net = torch.nn.Linear(in_features=1 ,out_features=1, bias=True) 

In [13]:
net.weight.data

tensor([[0.3480]])

In [14]:
net.weight.data=torch.tensor([[10.0]])

In [15]:
net.bias.data=torch.tensor([-5.0])

In [16]:
net.weight,net.bias

(Parameter containing:
 tensor([[10.]], requires_grad=True), Parameter containing:
 tensor([-5.], requires_grad=True))

In [17]:
net(x.reshape(100,1))
# input 차원을 nx1로 설정했기 때문에 차원이 nx2인 X는 사용할 수 없다.
# 이전 포스트 마지막 부분에서 언급한 shape 문제 때문에 X 또는 x 를 바로 사용할 수 없다.
# ( x는 그냥 크기가 100인 벡터인데 torch에서 요구하는 차원은 100x1 이므로 )
# 결과값이 100x1 의 행렬로 출력됨.

tensor([[-29.8211],
        [-28.6215],
        [-24.9730],
        [-21.2394],
        [-19.7919],
        [-19.6354],
        [-19.5093],
        [-19.4352],
        [-18.7223],
        [-18.0793],
        [-16.9040],
        [-16.0918],
        [-16.0536],
        [-15.8746],
        [-14.4690],
        [-14.3193],
        [-13.6426],
        [-12.8578],
        [-12.5486],
        [-12.4213],
        [-11.9484],
        [-11.1034],
        [-10.8296],
        [-10.6210],
        [-10.5064],
        [-10.0578],
        [ -9.8063],
        [ -9.7380],
        [ -9.7097],
        [ -9.6756],
        [ -8.8736],
        [ -8.7195],
        [ -8.6880],
        [ -8.1592],
        [ -7.7752],
        [ -7.7716],
        [ -7.7339],
        [ -7.7208],
        [ -7.6677],
        [ -7.1551],
        [ -7.0004],
        [ -6.8163],
        [ -6.7081],
        [ -6.5655],
        [ -6.4480],
        [ -6.3612],
        [ -6.0566],
        [ -5.6031],
        [ -5.5589],
        [ -5.2137],


#### 

### step2: loss

#### 방법1: 손실함수를 직접정의하는 방법

In [18]:
loss=torch.mean((y-yhat1)**2)
loss

tensor(85.8769, grad_fn=<MeanBackward0>)

In [19]:
loss=torch.mean((y-yhat2)**2)
loss

tensor(176.2661, grad_fn=<MeanBackward0>)

- 176.2661? 이건 잘못된 결과임

원래 두 결과값이 같아야 정상

이는 y는 길이가 100인 벡터인데 yhat2는 100x1인 행렬이라 생기는 문제

In [20]:
loss=torch.mean((y.reshape(100,1)-yhat2)**2)
loss

tensor(85.8769, grad_fn=<MeanBackward0>)

#### 방법2: torch.nn.MSELoss()를 사용하여 손실함수를 정의하는 방법 

In [21]:
lossfn=torch.nn.MSELoss()
# MSE 를 사용한 loss 값을 구해주는 함수 

In [22]:
loss=lossfn(y,yhat1)
loss

tensor(85.8769, grad_fn=<MseLossBackward>)

In [23]:
loss=lossfn(y.reshape(100,1),yhat2)
loss
# 마찬가지로 shape을 맞춰줘야 한다.
# MSE Loss 를 이용했으므로 'grad_fn=<MseLossBackward>' 라는 미분 꼬리표가 붙는다.

tensor(85.8769, grad_fn=<MseLossBackward>)

### [숙제](https://ieilms.jbnu.ac.kr/)

`-` model: $y_i= w_0+w_1 x_{i1}+w_2 x_{i2} +\epsilon_i = 2.5 + 4x_{1i} + -2x_{2i}+\epsilon_i, \quad i=1,2,\dots,n$ 

In [ ]:
torch.manual_seed(43052)
n=100
ones= torch.ones(n)
x1,_ = torch.randn(n).sort()
x2,_ = torch.randn(n).sort()
X = torch.vstack([ones,x1,x2]).T
W = torch.tensor([2.5,4,-2])
ϵ = torch.randn(n)*0.5
y = X@W + ϵ
ytrue = X@W

`-` torch.nn.Linear() 를 이용하여 $\bf{\hat{W}}=\begin{bmatrix}1 \\ 1 \\ 1 \end{bmatrix}$ 에 대한 $\hat{y}$를 구하라. 